sky/provision/lambda_cloud/instance.py

```diff
@@ -0,0 +1,271 @@
+    while True:
+        instances = _filter_instances(cluster_name_on_cloud, ['active'])
+        ready_instance_cnt = 0
+        for instance_id, instance in instances.items():
+            if instance.get('status') == 'active':
+                ready_instance_cnt += 1
+        if ready_instance_cnt == config.count:
+            break
+
+        time.sleep(POLL_INTERVAL)
+
+    assert head_instance_id is not None, 'head_instance_id should not be None'
+    return common.ProvisionRecord(
+        provider_name='lambda',
+        cluster_name=cluster_name_on_cloud,
+        region=region,
+        zone=None,
+        head_instance_id=head_instance_id,
+        resumed_instance_ids=[],
+        created_instance_ids=created_instance_ids,
+    )
+
+
+def wait_instances(region: str, cluster_name_on_cloud: str,
+                   state: Optional[status_lib.ClusterStatus]) -> None:
+    del region, cluster_name_on_cloud, state  # Unused.
+
+
+def stop_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    raise NotImplementedError(
+        'stop_instances is not supported for Lambda Cloud')
+
+
+def terminate_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    """See sky/provision/__init__.py"""
+    del provider_config
+    lambda_client = _get_lambda_client()
+    instances = _filter_instances(cluster_name_on_cloud, None)
+    # TODO (kmushegi): terminate all instances together since
+    # remove_instances accepts multiple instance ids
+    for instance_id, instance in instances.items():
+        logger.debug(f'Terminating instance {instance_id}: {instance}')
+        if worker_only and not instance['name'].endswith('-worker'):
+            continue
+        try:
+            lambda_client.remove_instances(instance_id)
+            logger.info(f'Terminated instance {instance_id}')
+        except Exception as e:  # pylint: disable=broad-except
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    f'Failed to terminate instance {instance_id}: '
+                    f'{common_utils.format_exception(e, use_bracket=False)}'
+                ) from e
+
+
+def get_cluster_info(
+    region: str,
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> common.ClusterInfo:
+    del region  # unused
+    running_instances = _filter_instances(cluster_name_on_cloud, ['active'])
+    instances: Dict[str, List[common.InstanceInfo]] = {}
+    head_instance_id = None
+    for instance_id, instance_info in running_instances.items():
+        instances[instance_id] = [
+            common.InstanceInfo(
+                instance_id=instance_id,
+                # TODO (kmushegi): check if this is correct;
+                # external ip is preferred
+                internal_ip='',
```

Comment:
Could we also check if is there a private IP for Lambda Cloud? From the previous implementation seems like they have:

https://github.com/skypilot-org/skypilot/blob/c7ee6a22c2fbb912bbe8f83b2931062e6b4e4490/sky/skylet/providers/lambda_cloud/node_provider.py#L232-L247

Reply:
I removed this somewhat purposefully since this check to break intermittently. it relies on the `docker0` network interface to be up and on some instances it is not up.

would it make sense to make a best effort to grab the internal ip? lmk

Reply:
I think the private ip would be preferred by some user who concerned about privacy; cc @Michaelvll for inputs here

Reply:
Our previous implementation for getting private IP is indeed quite hacky. Just wondering if Lambda has updated their API to include private IP? It seems to me that they do have it in their doc: https://cloud.lambdalabs.com/api/v1/docs#operation/listInstances

Maybe we should use that to see if we can connect our underlying ray cluster through that private IP?

Reply:
Ohhh word, great!

Reply:
this worked and fixed the multi node startup issues as well :)